File: src/onedimensiondp/lc_70_climbing_stairs/climbing_stairs.java
class Solution {

    Map<Integer, Integer> map = new HashMap<>();

    public int climbStairs(int n) {
        return dp(0, n);
    }

    public int dp(int steps, int n) {
        if (map.containsKey(steps))
            return map.get(steps);
        if (steps > n)
            return 0;
        if (steps == n)
            return 1;
        int ans = dp(steps + 1, n) + dp(steps + 2, n);
        map.put(steps, ans);
        return ans;
    }
}

File: src/onedimensiondp/lc_70_climbing_stairs/climbing_stairs.py
class Solution:
    def climbStairs(self, n: int) -> int:
        def rec(n, memo):
            if n in memo:
                return memo[n]
            memo[n] = rec(n-1,memo) + rec(n-2,memo)
            return memo[n]
        memo = {}
        memo[0] = 1
        memo[1] = 1
        return rec(n, memo)